source/geod24/concurrency.d

```diff
@@ -1006,6 +1007,601 @@ protected:
+
+                If a writer is already blocked on the Channel, reader copies the
+                value to `output` and wakes up the writer by yielding
+
+                If no writer is ready in the wait queue, reader appends itself
+                to read wait queue and blocks
+
+                If channel is closed, it returns immediatly with a failure
+
+            Buffered mode:
+
+                If there are existing messages in the buffer, reader pops one off
+                the buffer and returns immediatly with success, regardless of the
+                Channel being closed or not
+
+                If there are no messages in the buffer it behaves exactly like the
+                unbuffered operation
+
+            Param:
+                output = Reference to output variable
+            Returns:
+                Success/Failure
+
+        ***********************************************************************/
+
+        bool read(ref T output)
+        {
+            bool success = tryRead(output);
+
+            if (!success && !this.closed)
+            {
+                ChannelQueueEntry q_ent = enqueueEntry(this.readq, &output);
+
+                q_ent.sem.wait();
+                return q_ent.success;
+            }
+
+            return success;
+        }
+
+        /***********************************************************************
+
+            Try to read a message from Channel without blocking
+
+            tryRead locks the Channel and reads a message if it is possible
+            to do so without blocking. If the read would block, it returns without
+            unlocking the Channel to allow caller to `enqueueEntry` without competing
+            for the lock again.
+
+            If the tryRead is being executed in a select call, it tries to consume the
+            `caller_sel` with the given `caller_sel_id`. It only proceeds if it can
+            successfully consume the `caller_sel`
+
+            Param:
+                output = Field to write the message to
+                caller_sel = SelectState instace of the select call being executed
+                caller_sel_id = id of the select call being executed
+
+            Returns:
+                Success/Failure
+
+        ***********************************************************************/
+
+        private bool tryRead(ref T output, SelectState caller_sel = null, uint caller_sel_id = 0)
+        {
+            this.lock.lock();
+
+            ChannelQueueEntry write_ent = dequeueEntry(this.writeq, caller_sel, caller_sel_id);
+
+            if (this.buffer.length > 0)
+            {
+                // if dequeueEntry fails, we will try to consume caller_sel again.
+                if (!caller_sel || write_ent || caller_sel.tryConsume(caller_sel_id))
+                {
+                    output = this.buffer[0];
+                    this.buffer = this.buffer.remove(0);
+
+                    if (write_ent)
+                    {
```

Comment:
```suggestion
                    {
                        assert(write_ent.pVal !is null);
```

Reply:
Maybe this should be done at the enqueue stage to better localize the source.